feat(KeyboardShortcuts): add customizable shortcuts for common actions

maxpatiiuk opened the pull request:
Begins to fix: #1746

### Checklist

- [ ] Self-review the PR after opening it to make sure the changes look good
      and self-explanatory (or properly documented)
- [x] Add automated tests
- [x] Add relevant issue to release milestone

### Testing instructions

- Added `Keyboard Shortcuts` item to User tools - verify it looks good
- For each shortcut on the `Keyboard Shortcuts` page that has a default value assigned, make sure that the default keyboard shortcuts we have set do not already have an action assigned to them (we don't want to overwrite default browser or OS actions, unless we really have to, or unless user would expect us to do so)
  - i.e open any non-specify webpage (google.com) and see if this key does anything (this needs to be verified in non-specify pages because if specify collides with a system shortcut, specify will take priority, and so you may not even know that it collides)
  - Verify this on Windows in Chrome and Firefox
  - Verify this on macOS in Chrome and Safari
  - (if possible) Verify this on Linux in Firefox
- Check if any other common action is missing a keyboard shortcut that you believe we should add
- Verify the ability to set a keyboard shortcut to any page (see https://github.com/specify/specify7/pull/5097#issuecomment-2562140320)

<!-- What are the steps to verify the fixes/changes in this PR? -->
<!-- Can part of that be replaced by automated tests? -->

maxpatiiuk committed:
feat(KeyboardShortcuts): initial commit

maxpatiiuk committed:
feat(KeyboardShortcuts): complete base features

maxpatiiuk committed:
refactor(KeyboardShortcuts): move code into separate folder

maxpatiiuk committed:
refactor(KeyboardShortcuts): use key codes over key names

As per https://github.com/specify/specify7/issues/1746#issuecomment-2227113839

maxpatiiuk committed:
refactor(KeyboardShortcuts): rename from "Keyboard"

"Keyboard" is a bit too generic/ambiguous

maxpatiiuk committed:
fix(KeyboardShortcuts): polish UI

maxpatiiuk committed:
feat(KeyboardShortcuts): begin integrating with UI

maxpatiiuk committed:
fix(KeyboardShortcuts): correctly localize shortcuts on windows

maxpatiiuk committed:
feat(TreeView): improve layout

- Replace "show nodes with children only" checkbox with a toggle button.
  Reasons:
  - The checkbox was the only thing below the tree viewer - occupying
    an entire line of horizontal space.
  - Because of how Tab and Shift+Tab keys are used in the tree viewer,
    moving focus to elements below the tree viewer is not possible - you
    can only move above the tree viewer. That made this checkbox
    inaccessible from keyboard
  - The label for this checkbox is long, adding visual noise - it's
    cleaner as a small tidy button
- Don't display split view controls if split view is not enabled to
  reduce visual clutter

maxpatiiuk committed:
feat(KeyboardShortcuts): Add shortcuts for forms, tree and queries

grantfitzsimmons commented:
CSIRO mentioned users are having repetitive strain injuries due to this not being implemented

maxpatiiuk commented:
Bugs:
- Can add same shortcut twice for the same action
- When adding the keyboard shortcut, can't reach the "Remove" button with keyboard alone

maxpatiiuk committed:
fix(KeyboardShortcuts): prevent adding duplicate shortcuts

maxpatiiuk committed:
fix(KeyboardShortcuts): don't permit binding to Shift+Tab

maxpatiiuk committed:
feat(KeyboardShortcuts): add custom URL shortcuts

maxpatiiuk commented:
Added ability to set shortcut to any URL:

https://github.com/user-attachments/assets/48a445d9-af3a-4332-ad55-53ccb1191221

Use cases:
- can set a keyboard shortcut to any menu item/user tool
- can set a keyboard shortcut to a record set/query that you use often
- can set a keyboard shortcut to an external website (morphosource/gbif/...)
  - this was trivial to implement, and makes the feature much more powerful - will see how it will end up being used
- brings us closer to https://github.com/specify/specify7/discussions/5686

By default, I only included keyboard shortcut to open the keyboard shortcut viewer, and will add shortcut to go to the home page. For the rest, we can let people set those to what makes sense to them (I avoided setting too many be default as things can conflict with os/browser/browser extensions, unless we go for 3-4 key shortcuts)

TODO:
- [ ] come up with a default keyboard shortcut to go to the home page

emenslin reviewed (changes requested):
Can we make the save button orange or something in the shortcuts to url dialog? I think if you're editing a url it can be easy to not click the save button assuming that when you close out of the dialog the bigger save button will be orange. Or we could also consider having some sort of blocker or dialog if that would make more sense. I feel like when first testing the dialog I didn't even notice the save button appeared next to it but that could just be me. 

https://github.com/user-attachments/assets/d47d1ecb-6829-447d-8cd0-d14e9f7dc968


On top of that if you create a new shortcut and don't press the gray save button, close the dialog, the save button on the naviagation menu turns orange but if you click that and go back to the shortcut it says 'No key binding assigned'

https://github.com/user-attachments/assets/51598069-891c-490a-9261-e186e4907622


If you add a shortcut and then try to add another without saving the dialog freaks out 

https://github.com/user-attachments/assets/8dbbdb6c-f7cd-411f-9040-8a1d68101294



Also I feel like this layout can be a little confusing with the buttons and the spacing. Is the intention of the add button after a shortcut has been made for people to be able to add multiple shortcuts to one URL? And if that is the case could they all be in one line instead of having the add button below the shortcut and add button? 
![image](https://github.com/user-attachments/assets/f4c48af8-3c69-432f-90d7-242d23a856f2)


Also is there a plan to add more to this in the future? I feel like it looking just like preferences but only having the keyboard shortcuts looks a little odd especially considering the keyboard shortcuts opens a separate dialog so it just feels a little bare and confusing. 
![image](https://github.com/user-attachments/assets/d8db31dc-61b9-4050-a752-d7cc813cf4fb)

CarolineDenis commented:
@maxpatiiuk, is this PR waiting on final testing or there are still some code changes needed?

maxpatiiuk committed:
fix(KeyboardShortcuts): respond to review feedback

maxpatiiuk commented:
> Can we make the save button orange or something in the shortcuts to url dialog?

Great point. Fixed

> On top of that if you create a new shortcut and don't press the gray save button, close the dialog, the save button on the naviagation menu turns orange

Fixed. The button no longer turns orange unless you hit "Save" inside the dialog

> If you add a shortcut and then try to add another without saving the dialog freaks out

Ah, geez. Fixed!

Now, as soon as you start editing any keyboard shortcut, all others become disabled - leaving you with only 2 options - complete entering the shortcut and hit "Save" or hit "Remove"

<img width="1033" height="634" alt="image" src="https://github.com/user-attachments/assets/dd6db9a4-4a9e-44a1-bfea-b6e5a9f2e07a" />

> Also is there a plan to add more to this in the future? I feel like it looking just like preferences but only having the keyboard shortcuts looks a little odd especially considering the keyboard shortcuts opens a separate dialog so it just feels a little bare and confusing.

Hmm, not sure now you got the page to look this way.

The keyboard shortcuts page should look more complete:

![Screenshot 2025-07-26 at 08 51 40](https://github.com/user-attachments/assets/5e51b6c2-2ea7-4bf0-9733-a11aa9d982bf)

It is possible that the page failed to render after this bug happened:
> If you add a shortcut and then try to add another without saving the dialog freaks out


---

Merge conflicts resolved. Ready for testing.

maxpatiiuk committed:
refactor(QueryBuilder): self-review code

maxpatiiuk committed:
docs(KeyboardShortcuts): add instructions on adding new shortcuts

maxpatiiuk committed:
build(KeyboardShortcuts): avoid cyclical import

maxpatiiuk committed:
tests(Preferences): fix localization tests
